File: app/src/main/java/com/robotnec/budget/app/adapters/support/HeaderAdapter.java
package com.robotnec.budget.app.adapters.support;

import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

/**
 * @author zak <zak@swingpulse.com>
 */
public class HeaderAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private final static int VIEW_TYPE_HEADER = 1;
    private final static int VIEW_TYPE_BODY = 2;

    private final RecyclerView.Adapter body;
    private final ListHeader listHeader;

    public HeaderAdapter(RecyclerView.Adapter body, ListHeader listHeader) {
        this.body = body;
        this.listHeader = listHeader;
        setHasStableIds(true);
    }

    @Override
    public long getItemId(int position) {
        if (position == 0) {
            return -1;
        } else {
            return body.getItemId(position - 1);
        }
    }

    @Override
    public int getItemViewType(int position) {
        if (position == 0) {
            return ViewTypeHash.hash(VIEW_TYPE_HEADER, 0);
        } else {
            return ViewTypeHash.hash(VIEW_TYPE_BODY, body.getItemViewType(position - 1));
        }
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewTypeHash) {
        if (ViewTypeHash.getAdapterViewType(viewTypeHash) == VIEW_TYPE_HEADER) {
            return listHeader.onCreateViewHolder(parent);
        } else {
            return body.onCreateViewHolder(parent, ViewTypeHash.getItemViewType(viewTypeHash));
        }
    }

    @SuppressWarnings("unchecked")
    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        if (position == 0) {
            listHeader.onBindViewHolder(holder);
        } else {
            body.onBindViewHolder(holder, position - 1);
        }
    }

    @Override
    public int getItemCount() {
        return body.getItemCount() + 1;
    }

    private static class ViewTypeHash {
        static int hash(int adapterViewType, int itemViewType) {
            return adapterViewType << 8 | itemViewType;
        }

        static int getAdapterViewType(int hash) {
            return hash >> 8;
        }

        static int getItemViewType(int hash) {
            return hash & 0xFF;
        }
    }
}

File: app/src/main/java/com/robotnec/budget/app/adapters/support/HeaderAdapter.kt
package com.robotnec.budget.app.adapters.support

import androidx.recyclerview.widget.RecyclerView
import android.view.ViewGroup

/**
 * @author zak zak@swingpulse.com>
 */
class HeaderAdapter(private val body: RecyclerView.Adapter<RecyclerView.ViewHolder>,
                    private val listHeader: ListHeader) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    init {
        setHasStableIds(true)
    }

    override fun getItemId(position: Int): Long {
        return if (position == 0) {
            -1
        } else {
            body.getItemId(position - 1)
        }
    }

    override fun getItemViewType(position: Int): Int {
        return if (position == 0) {
            ViewTypeHash.hash(VIEW_TYPE_HEADER, 0)
        } else {
            ViewTypeHash.hash(VIEW_TYPE_BODY, body.getItemViewType(position - 1))
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewTypeHash: Int): RecyclerView.ViewHolder {
        return if (ViewTypeHash.getAdapterViewType(viewTypeHash) == VIEW_TYPE_HEADER) {
            listHeader.onCreateViewHolder(parent)
        } else {
            body.onCreateViewHolder(parent, ViewTypeHash.getItemViewType(viewTypeHash))
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (position == 0) {
            listHeader.onBindViewHolder(holder)
        } else {
            body.onBindViewHolder(holder, position - 1)
        }
    }

    override fun getItemCount(): Int {
        return body.itemCount + 1
    }

    private object ViewTypeHash {
        internal fun hash(adapterViewType: Int, itemViewType: Int): Int {
            return adapterViewType shl 8 or itemViewType
        }

        internal fun getAdapterViewType(hash: Int): Int {
            return hash shr 8
        }

        internal fun getItemViewType(hash: Int): Int {
            return hash and 0xFF
        }
    }

    companion object {
        private const val VIEW_TYPE_HEADER = 1
        private const val VIEW_TYPE_BODY = 2
    }
}
